components/calendar/src/chinese_based.rs

```diff
@@ -309,6 +344,103 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
         result
     }
+
+    /// Create a new arithmetic date from a year, month ordinal, and day with bounds checking; returns the
+    /// result of creating this arithmetic date, as well as a ChineseBasedCache - either the one passed in
+    /// optionally as an argument, or a new ChineseBasedCache for the given year, month, and day args.
+    pub(crate) fn new_from_ordinals(
+        year: i32,
+        month: u8,
+        day: u8,
+        cache: &ChineseBasedCache,
+    ) -> Result<ArithmeticDate<C>, CalendarError> {
+        let max_month = Self::months_in_year_cached(cache);
+        if month > max_month {
+            return Err(CalendarError::Overflow {
+                field: "month",
+                max: max_month as usize,
+            });
+        }
+
+        let max_day = Self::days_in_month_cached(month, cache);
+        if day > max_day {
+            return Err(CalendarError::Overflow {
+                field: "day",
+                max: max_day as usize,
+            });
+        }
+
+        // Unchecked can be used because month and day are already checked in this fn
+
+        Ok(ArithmeticDate::<C>::new_unchecked(year, month, day))
+    }
+
+    /// Call `months_in_year_cached` on a `ChineseBasedDateInner`
+    pub(crate) fn months_in_year_inner(&self) -> u8 {
+        Self::months_in_year_cached(&self.1)
+    }
+
+    /// Return the number of months in a given year, which is 13 in a leap year, and 12 in a common year.
+    /// Also takes a `ChineseBasedCache` argument.
+    fn months_in_year_cached(cache: &ChineseBasedCache) -> u8 {
+        if cache.is_leap_year {
+            13
+        } else {
+            12
+        }
+    }
+
+    /// Calls `days_in_month_cached` on an instance of ChineseBasedDateInner
+    pub(crate) fn days_in_month_inner(&self) -> u8 {
+        Self::days_in_month_cached(self.0.month, &self.1)
+    }
+
+    /// Returns the number of days in the given `year`, `month` (see `month_days`);
+    /// this function takes a `ChineseBasedCache` argument.
+    fn days_in_month_cached(month: u8, cache: &ChineseBasedCache) -> u8 {
+        let new_year = cache.new_year;
+        let approx = new_year + ((month - 1) as i64 * 29);
+        let prev_new_moon = Self::new_moon_before((approx + 15).as_moment());
+        let next_new_moon = Self::new_moon_on_or_after((approx + 15).as_moment());
+        let result = (next_new_moon - prev_new_moon) as u8;
+        debug_assert!(result == 29 || result == 30);
+        result
+    }
+
+    /// Calls day_in_year_cached on an instance of ChineseBasedDateInner
+    pub(crate) fn days_in_year_inner(&self) -> u16 {
+        Self::days_in_year_cached(&self.1)
+    }
+
+    /// Returns the number of day in the given year with an associated `ChineseBasedCache`.
+    fn days_in_year_cached(cache: &ChineseBasedCache) -> u16 {
+        let prev_new_year = cache.new_year;
+        let next_new_year = Self::new_year_on_or_before_fixed_date(prev_new_year + 370, None).0;
```

Comment:
Issue: This also isn't really "cached". Please try for a similar type of abstraction as suggested above. You already have a `days_in_provided_year` function; can you merge these? I don't like multiple functions that do the same thing, especially in complex code like this.

Thought: Can we cache the following new year? I think we need that anyway when computing whether we are in a leap year.

Reply:
Done for changing the function; I tried having the next new year included as part of the cache and in cache calculations, but for some reason this resulted in a significant decrease in performance (~49 -> 65 microseconds). There might be ways to do it that are better than what I did, but I don't think it's that crucial, so I'm not going to implement it now. (unless you really think it's necessary)

Reply:
The thing is that you are already computing `is_leap_year`, and in order to compute that, you need to compute both the prior new year and the next new year. If it got slower, it's probably because you were duplicating work.